### ISSUE TITLE
Outer type guards do not apply in nested functions

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#1731]
**Describe the bug**
Type guards evaluated before nested function calls are ignored by the type checker.

**To Reproduce**
Create a nested function with a typing guard in the outer function.

```py
from typing import Optional, Union

def foo(item: Union[str, int]) -> Optional[str]:
    if isinstance(item, int):
        return None
    
    def bar() -> str:
        return item.capitalize()

    return bar()
```

**Expected behavior**
in the inner function `bar` variable `item` would be assumed to be of type `str` rather than `str | int`

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/8677174/113560570-877e2380-9646-11eb-8a4d-13030ed90290.png)

**VS Code extension or command-line**
VS Code extension. Pylance version 2021.3.4

**Additional context**
mypy as of its latest release (v0.812) [also suffers from this issue](/python/mypy/issues/10280).

**Comment by erictraut** [comment]
This is intended behavior. See https://github.com/microsoft/pyright/issues/599 for a full explanation.

In summary, it's not safe to apply type narrowing to a captured variable that's used in a nested function or lambda because the value (and therefore the type) could be changed prior to the function or lambda being invoked. The solution is to assign the narrowed value to a new local variable that has the narrowed type.

```python
item_str = item

def bar() -> str:
    return item_str.capitalize()
```

**Comment by erictraut** [comment]
Reopening this issue and marking it fixed. After thinking about it more, I found a way to detect when it's safe to apply type narrowing for captured local variables and parameters. See https://github.com/microsoft/pyright/issues/2343 for details.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.172, which I just published. It will also be in the next release of pylance.
